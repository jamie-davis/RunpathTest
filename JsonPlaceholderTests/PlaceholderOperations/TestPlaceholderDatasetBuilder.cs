using System.Collections.Generic;
using System.Linq;
using JsonPlaceholder.PlaceholderOperations;
using JsonPlaceholderTests.Data;
using JsonPlaceholderTests.TestUtils;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Newtonsoft.Json.Linq;
using TestConsoleLib.Testing;
using Xunit;

namespace JsonPlaceholderTests.PlaceholderOperations
{
    public class TestPlaceholderDatasetBuilder
    {
        private PlaceholderDatasetCombiner _combiner;
        private ILogger<PlaceholderDatasetBuilder> _logger;
        private FakeLog _fakeLog;
        private FakePlaceholderFetcher _fetcher;

        public TestPlaceholderDatasetBuilder()
        {
            //Trim the test data
            IEnumerable<JObject> AlbumFilter(IEnumerable<JObject> albums)
            {
                return albums.GroupBy(a => a["userId"])
                    .Take(2) //only take two users
                    .SelectMany(g => g.Take(2)); //only take the first two albums for each user
            }

            IEnumerable<JObject> PhotoFilter(IEnumerable<JObject> photos)
            {
                return photos.GroupBy(p => p["albumId"])
                    .SelectMany(g => g.Take(2)); //Take two photos from each album
            }

            //The test data has been trimmed to just the first 2 photos in each album, and the albums have been trimmed to 2 users with 2 albums each.
            _fetcher = new FakePlaceholderFetcher(AlbumFilter, PhotoFilter);

            _fakeLog = new FakeLog(l => true);
            _logger = new FakeLogger<PlaceholderDatasetBuilder>(_fakeLog, "Test");
        }

        [Fact]
        public void FormatAllData()
        {
            //Arrange
            var builder = new PlaceholderDatasetBuilder(_fetcher, _logger);

            //Act
            var result = builder.BuildAsync().Result; //data is limited to 2 users, 2 albums each with 2 photos in each album, but we are selecting them all

            //Assert
            JArray.FromObject(result).ToString().Verify(); // simulate what the JsonResult will do and verify the output
        }

        [Fact]
        public void FormatAllDataForUser()
        {
            //Arrange
            var builder = new PlaceholderDatasetBuilder(_fetcher, _logger);

            //Act
            var result = builder.BuildAsync(2).Result; //data is limited to 2 users, 2 albums each with 2 photos in each album, we are only selecting one user.

            //Assert
            JArray.FromObject(result).ToString().Verify(); // simulate what the JsonResult will do and verify the output
        }
    }
}
