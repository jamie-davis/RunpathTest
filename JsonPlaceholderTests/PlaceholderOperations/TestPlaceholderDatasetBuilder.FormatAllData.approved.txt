[
  {
    "UserId": 1,
    "Id": 1,
    "Title": "quidem molestiae enim",
    "Photos": [
      {
        "Id": 1,
        "Title": "accusamus beatae ad facilis cum similique qui sunt",
        "Uri": "https://via.placeholder.com/600/92c952",
        "ThumbnailUri": "https://via.placeholder.com/150/92c952"
      },
      {
        "Id": 2,
        "Title": "reprehenderit est deserunt velit ipsam",
        "Uri": "https://via.placeholder.com/600/771796",
        "ThumbnailUri": "https://via.placeholder.com/150/771796"
      }
    ]
  },
  {
    "UserId": 1,
    "Id": 2,
    "Title": "sunt qui excepturi placeat culpa",
    "Photos": [
      {
        "Id": 51,
        "Title": "non sunt voluptatem placeat consequuntur rem incidunt",
        "Uri": "https://via.placeholder.com/600/8e973b",
        "ThumbnailUri": "https://via.placeholder.com/150/8e973b"
      },
      {
        "Id": 52,
        "Title": "eveniet pariatur quia nobis reiciendis laboriosam ea",
        "Uri": "https://via.placeholder.com/600/121fa4",
        "ThumbnailUri": "https://via.placeholder.com/150/121fa4"
      }
    ]
  },
  {
    "UserId": 2,
    "Id": 11,
    "Title": "quam nostrum impedit mollitia quod et dolor",
    "Photos": [
      {
        "Id": 501,
        "Title": "asperiores exercitationem voluptates qui amet quae necessitatibus facere",
        "Uri": "https://via.placeholder.com/600/cda4c0",
        "ThumbnailUri": "https://via.placeholder.com/150/cda4c0"
      },
      {
        "Id": 502,
        "Title": "omnis qui sit et",
        "Uri": "https://via.placeholder.com/600/74e371",
        "ThumbnailUri": "https://via.placeholder.com/150/74e371"
      }
    ]
  },
  {
    "UserId": 2,
    "Id": 12,
    "Title": "consequatur autem doloribus natus consectetur",
    "Photos": [
      {
        "Id": 551,
        "Title": "eveniet debitis nihil",
        "Uri": "https://via.placeholder.com/600/21e334",
        "ThumbnailUri": "https://via.placeholder.com/150/21e334"
      },
      {
        "Id": 552,
        "Title": "odit culpa optio nesciunt",
        "Uri": "https://via.placeholder.com/600/b56655",
        "ThumbnailUri": "https://via.placeholder.com/150/b56655"
      }
    ]
  }
]